templates/netlify/remix.init/vscode.json

```diff
@@ -0,0 +1,12 @@
+{
+  "deno.enablePaths": [
+    "server.js",
+    "server.ts",
+    "globals.d.ts",
+    "netlify/edge-functions/",
+    ".netlify/edge-functions/"
+  ],
+  "deno.unstable": true,
+  "deno.importMap": ".netlify/edge-functions-import-map.json",
```

Comment:
This file doesn't seem to exist?

We should probably have a `.vscode` folder instead (like we have in `deno` template) and copy over the whole folder if people choose for Netlify Edge

Reply:
remix.init generates a `.vscode` folder for this. Also, you mentioned it elsewhere, but we don't need the import map like the deno server build target. We only need our import map because it's only for Netlify Edge functions.

```javascript
 const files = [
   ...
   ["vscode.json", join(".vscode", "settings.json")],
   ...
 ];

async function main({ rootDirectory }) {
  if (await shouldUseEdge()) {
    await fs.mkdir(join(rootDirectory, ".vscode"));

    for (let [file, target] of files) {
      await fs.copyFile(
        join(rootDirectory, "remix.init", file),
        join(rootDirectory, target || file)
      );
    }
  }
}
```

https://github.com/nickytonline/remix/blob/d65f070c5f38313abf6ef4d9599d468400a31b50/templates/netlify/remix.init/index.js#L10

Reply:
I understand `vscode.json` is copied to `.vscode/settings.json` when people choose for Netlify Edge in `remix.init` & that's how I would expect it to be.

However, it is referencing `edge-functions-import-map.json`, which doesn't exist in the template.

I would suggest to create `remix.init/vscode` with a `settings.json` & `resolve_npm_imports.json` file in it & copy the whole folder to `.vscode` when people choose for Netlify Edge.

Reply:
I forgot to mention `edge-functions-import-map.json` is [generated by us](https://answers.netlify.com/t/edge-functions-import-map-json/55298) once they've set up the template and build the first time. 

I'm probably not explaining it well but we don't need the imports that the deno server build target has for the VS Code deno extension.

Reply:
Will it also include all used dependencies, like we have in the `deno` template?

```json
"imports": {
  "@remix-run/deno": "https://esm.sh/@remix-run/deno@1.5.0",
  "@remix-run/dev/server-build": "https://esm.sh/@remix-run/dev@1.5.0/server-build",
  "@remix-run/react": "https://esm.sh/@remix-run/react@1.5.0",
  "react": "https://esm.sh/react@17.0.2",
  "react-dom": "https://esm.sh/react-dom@17.0.2",
  "react-dom/server": "https://esm.sh/react-dom@17.0.2/server"
}
```

Reply:
It doesn't, but as @ascorbic explained to me, we don’t need those because the deno extension won’t be enabled for the Remix code, it's just for our Edge functions.

Reply:
@pcattori If that's the case, why do we have the import map in the `deno` template? 🤔

Either it's not the case & we should figure out how to merge `edge-functions-import-map.json` with the one from the template _or_ we should remove the imports map from the `deno` template as it's not used.

Reply:
We expect users of the `deno` template to enable the Deno extensions for VS Code for the whole project, but we use NPM-managed dependencies, so we need to provide types for the Deno extension via the import map.

See https://github.com/remix-run/remix/blob/main/decisions/0001-use-npm-to-manage-npm-dependencies-for-deno-projects.md#vs-code-type-hints .

If you are not expecting users to use the Deno extension for the Remix code, then VS Code will correctly resolve bare imports (i.e. not URL imports) like `import {...} from "@remix-run/react"` by looking at `node_modules/` and getting types from there.

Reply:
> and getting types from there

This means we need to have `@types/react` & `@types/react-dom` in `devDependencies`, right?

Nevertheless, I think we should keep things consistent/streamlined & expect people to install the Deno extension when developing with this new template as well.

Reply:
We do expect users to have the extension installed, but the VS Code settings that we generate use enablePaths to ensure it's only enabled for the edge functions themselves, and not for any Remix code (because that's compiled before Deno sees it). This is what the Netlify CLI offers to do for any site that runs edge functions locally. It's still rare that people are using Deno to build their sites, so we don't enable the Deno extension for that code.

Reply:
@pcattori if you're expecting users to enable it for the whole project, how does it deal with syntax that is incompatible between Remix and Deno, like `.ts` imports?

Reply:
@ascorbic : that's a good point! We want to allow `.ts` extension for relative imports.

Which reminds me: I need to write a quick `esbuild` plugin that resolves relative imports ending in `.ts` so that `esbuild` find those files too. Bug fix incoming!

Reply:
Nevermind, our compiler resolves `.{ts,tsx,js,jsx}` extensions in imports already. So we recommend using `.ts` and `.tsx` extensions for Remix + Deno projects. Both the Deno extension and the Remix compiler will work with that.

Reply:
@MichaelDeBoey, `.netlify/edge-functions-import-map.json` gets generated by us when we create the Edge function.

Reply:
@nickytonline I assume it won't contain the other imports that we have in `templates/deno/.vscode/resolve_npm_imports.json`?
https://github.com/remix-run/remix/blob/2248669ed59fd716e267ea41df5d665d4781f4a9/templates/deno/.vscode/resolve_npm_imports.json#L6-L15

Reply:
@MichaelDeBoey, all it contains is 

```json
{"imports":{"netlify:edge":"https://edge.netlify.com/v1/index.ts"}}
```

![CleanShot 2022-11-25 at 16 35 31](https://user-images.githubusercontent.com/833231/204057372-052055f9-380e-4392-b97e-ab7ae1783a1a.png)

Reply:
That's indeed as I expected.

Is there a way to point VS Code & Deno to both files?
We want to make sure that the ones we have in the Deno template (which we should have in the Netlify Edge template as well) are recognized as well, so we have automatic typings for React/...

Reply:
@nickytonline I think the best solution here would be to use whatever is in `.vscode/resolve_npm_imports.json` of the deno template & add the `netlify:edge` line tbh 🤔

Reply:
Unlike the Deno template, we don't treat any of the site files as Deno files, so we don't need the import map for those. This is why `enablePaths` doesn't include any site source files. As they're transpiled by Remix, we can treat them as node files and they'll pick up the typings from node_modules. We don't actually need the `netlify:edge` alias either, because we recommend using `https://edge.netlify.com` instead now. It's just being generated for backward-compatibility.

Reply:
So if I understand correctly, the template should look/work like a Node template, except using `@remix-run/deno` instead of `@remix-run/node`?

Reply:
@ascorbic So if I understand correctly, the template should look/work like a Node template, except using `@remix-run/deno` instead of `@remix-run/node`?

Reply:
Pretty much, yes. It's the emitted code that needs to be Deno-compatible, not the source.

Reply:
You can do either. Remix compiles everything with Node, so it needs to be acceptable to rollup, but it passes URL imports through unchanged so you can use those too and Deno will use them. You can use Deno globals too.

Reply:
Then I think we should treat this template as the Deno template, which means we don't need a `tsconfig.json` & we need everything we have in the Deno template in `.vscode/resolve_npm_imports.json`

CC/ @pcattori